### ISSUE TITLE
Regression: RecursionError in cse since gh-24303

### ISSUE THREAD

**Issue by bjodah** [sympy/sympy#24818]
I got a recursion error after updating SymPy to `master` in a project of mine, I managed to minimize the expression to a minimal reproducer: `cse(-1-UnevaluatedExpr(Symbol('x')+1))`

Here's a shell session identifying gh-24303 as the point of regression (cc @sylee957):
<details>
<summary> <pre>$ git checkout 978de3ace373 </pre> </summary>

```shell-session
HEAD is now at 978de3ace3 Merge pull request #24303 from sylee957/evaluate
$ python -c "from sympy import cse, UnevaluatedExpr, Symbol; print(cse(-1-UnevaluatedExpr(Symbol('x')+1)))"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/bjorn/vc/sympy/sympy/simplify/cse_main.py", line 824, in cse
    replacements, reduced_exprs = tree_cse(reduced_exprs, symbols, opt_subs,
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/bjorn/vc/sympy/sympy/simplify/cse_main.py", line 694, in tree_cse
    reduced_e = _rebuild(e)
                ^^^^^^^^^^^
  File "/home/bjorn/vc/sympy/sympy/simplify/cse_main.py", line 668, in _rebuild
    new_args = list(map(_rebuild, args))
               ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/bjorn/vc/sympy/sympy/simplify/cse_main.py", line 668, in _rebuild
    new_args = list(map(_rebuild, args))
               ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/bjorn/vc/sympy/sympy/simplify/cse_main.py", line 668, in _rebuild
    new_args = list(map(_rebuild, args))
               ^^^^^^^^^^^^^^^^^^^^^^^^^
  [Previous line repeated 990 more times]
  File "/home/bjorn/vc/sympy/sympy/simplify/cse_main.py", line 657, in _rebuild
    if c == [1]:
       ^^^^^^^^
  File "/home/bjorn/vc/sympy/sympy/core/numbers.py", line 2243, in __eq__
    return (self.p == other)
            ^^^^^^^^^^^^^^^
RecursionError: maximum recursion depth exceeded in comparison
$ git checkout 978de3ace373~1
HEAD is now at c303e9e4d3 Merge pull request #24302 from eagleoflqj/flake8
$ python -c "from sympy import cse, UnevaluatedExpr, Symbol; print(cse(-1-UnevaluatedExpr(Symbol('x')+1)))"
([], [-(1 + (x + 1))])
```

</details>

This might do as a regression test for a future fix (untested of course so no guarantees):
<details>
<summary>Proposed regression test</summary>

```patch
diff --git a/sympy/simplify/tests/test_cse.py b/sympy/simplify/tests/test_cse.py
index df82df3b84..08213f6e47 100644
--- a/sympy/simplify/tests/test_cse.py
+++ b/sympy/simplify/tests/test_cse.py
@@ -5,6 +5,7 @@
 from sympy.codegen.matrix_nodes import MatrixSolve
 from sympy.core.add import Add
 from sympy.core.containers import Tuple
+from sympy.core.expr import UnevaluatedExpr
 from sympy.core.function import Function
 from sympy.core.mul import Mul
 from sympy.core.power import Pow
@@ -534,6 +535,18 @@ def test_cse_ignore_issue_15002():
     rl = [e.subs(reversed(substs)) for e in reduced]
     assert rl == l
 
+
+def test_cse_unevaluated():
+    xp1 = UnevaluatedExpr(x + 1)
+    (x0, ue), red = cse([(-1-xp1)/(1-xp1)])  # this used to cause RecursionError
+    if ue == xp1:
+        assert red == [(-1-x0)/(1-x0)]
+    elif ue == -xp1:
+        assert red == [(-1+x0)/(1+x0)]
+    else:
+        raise ValueError("Unexpected common subexpression found")
+
+
 def test_cse__performance():
     nexprs, nterms = 3, 20
     x = symbols('x:%d' % nterms)
```

</details>

**Comment by sylee957** [comment]
I guess the problem comes from negative sign not getting cancelled for `UnevaluatedExpr`
```
x = Symbol('x')
expr = -1 - UnevaluatedExpr(x + 1)
-expr
```

**Comment by sylee957** [comment]
I think that `UnevaluatedExpr.is_commutative` undefined is relevant for this problem though
I could add this to `UnevaluatedExpr` to resolve.
```
+    def _eval_is_commutative(self):
+        return self.args[0].is_commutative
```

**Comment by bjodah** [comment]
@sylee957 that defeats the purpose of `UnevaluatedExpr` as far as I have understood its mechanism: the fact that it does not propagate `is_commutative` is what protects it from automatically simplifying in some cases (but I might have misunderstood things here...). Note that cse used to work for this example.

**Comment by oscarbenjamin** [comment]
Having `is_commutative = False` is one thing but having `is_commutative = None` should be disallowed. Either the object commutes with everything and can be moved around in a Mul (`commutative = True`) or it does not.

**Comment by bjodah** [comment]
I actually think there is a bug in `_rebuild` in `tree_cse`, I'm stepping through the code (using my suggested test case) and this line looks suspicious:

https://github.com/sympy/sympy/blob/75691e5cbdb3dfda4c8164a217bac6a6434404f6/sympy/simplify/cse_main.py#L668-L670

From the debugger on L669:
```
>>> expr
-1 - 2*(x + 1)
>>> opt_subs[expr]
Uneval<<class 'sympy.core.mul.Mul'>>(-1, -(-1 - 2*(x + 1)))
```

**Comment by sylee957** [comment]
I'm just going to push the above fix to `cse` if there is no better idea to do this.
Unfortunately, if we just revert #24303 the corresponding issue raises again too.

**Comment by bjodah** [comment]
sure, let's do that. thank you @sylee957

**Comment by oscarbenjamin** [comment]
I guess once a fix is merged is should also be backported to the 1.12 branch. I'm going to set the milestone on this issue.

**Comment by oscarbenjamin** [comment]
Reopening until a fix is on the 1.12 branch.
